js/src/common/utils/ItemList.ts

```diff
@@ -81,51 +169,143 @@ export default class ItemList {
    * Remove an item from the list.
    */
   remove(key: string): this {
-    delete this.items[key];
+    delete this._items[key];
 
     return this;
   }
 
   /**
    * Merge another list's items into this one.
+   *
+   * The list passed to this function will overwrite items which already exist
+   * with the same key.
    */
-  merge(items: this): this {
-    for (const i in items.items) {
-      if (items.items.hasOwnProperty(i) && items.items[i] instanceof Item) {
-        this.items[i] = items.items[i];
+  merge<K>(otherList: ItemList<K>): ItemList<T | K> {
```

Comment:
Can't we require that the other item list has the same type as this one statically?

Reply:
Requiring that seems like a bad idea. Merging two ItemLists with different types seems like something that should be allowed.

Reply:
> Requiring that seems like a bad idea. Merging two ItemLists with different types seems like something that should be allowed.

My only concern is that we lose out on catching errors where incompatible ItemLists are merged by accident.

Idea: could we have the generic here be something like:

`merge<N extends K = K>(otherList: ItemList<N>): ItemList<N> {`

Then, by default, `merge(otherList)` would require `otherList` to be of `K`, but `merge<K|T>(otherList)` where `otherList \in ItemList<T> would work.

Reply:
Looking back on this now, I agree that we should just type that the ItemLists should be the same.

The correct type for the merged itemlist should be set when the itemlist is first initialised. We shouldn't be changing the type mid-way through, as that causes a multitude of problems.